File: crates/tethys-schema/src/sync.rs
//! Sync engine wire types (M1.4, `architecture.md` §11).
//!
//! These are the only sync shapes crossing the IPC boundary. `tethys-sync`
//! speaks them directly so the command contract cannot drift from the engine.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use specta::Type;

/// Stable workspace identifier.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Type)]
#[serde(transparent)]
pub struct WorkspaceId(pub String);

impl WorkspaceId {
    pub fn new(id: impl Into<String>) -> Self {
        Self(id.into())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl From<&str> for WorkspaceId {
    fn from(value: &str) -> Self {
        Self(value.to_string())
    }
}

impl From<String> for WorkspaceId {
    fn from(value: String) -> Self {
        Self(value)
    }
}

impl std::fmt::Display for WorkspaceId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl AsRef<str> for WorkspaceId {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

/// A file surface that a registry entry can be projected to.
#[derive(
    Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Type,
)]
#[serde(rename_all = "kebab-case")]
pub enum ProjectionTarget {
    ClaudeCode,
    Codex,
    OpenCode,
}

impl ProjectionTarget {
    pub const FILES: [ProjectionTarget; 3] = [
        ProjectionTarget::ClaudeCode,
        ProjectionTarget::Codex,
        ProjectionTarget::OpenCode,
    ];

    pub fn as_str(self) -> &'static str {
        match self {
            ProjectionTarget::ClaudeCode => "claude-code",
            ProjectionTarget::Codex => "codex",
            ProjectionTarget::OpenCode => "opencode",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "claude-code" => Some(ProjectionTarget::ClaudeCode),
            "codex" => Some(ProjectionTarget::Codex),
            "opencode" => Some(ProjectionTarget::OpenCode),
            _ => None,
        }
    }
}

/// A file surface or session that a registry entry can be projected to (legacy).
#[derive(
    Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Type,
)]
#[serde(rename_all = "kebab-case")]
pub enum TargetId {
    Session,
    ClaudeCode,
    Codex,
    OpenCode,
}

impl TargetId {
    /// Targets that own a config file (not the session handoff).
    pub const FILES: [TargetId; 3] = [TargetId::ClaudeCode, TargetId::Codex, TargetId::OpenCode];

    /// Wire name used in `x-tethys.targets`.
    pub fn as_str(self) -> &'static str {
        match self {
            TargetId::Session => "session",
            TargetId::ClaudeCode => "claude-code",
            TargetId::Codex => "codex",
            TargetId::OpenCode => "opencode",
        }
    }

    /// Parses the wire name.
    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "session" => Some(TargetId::Session),
            "claude-code" => Some(TargetId::ClaudeCode),
            "codex" => Some(TargetId::Codex),
            "opencode" => Some(TargetId::OpenCode),
            _ => None,
        }
    }
}

impl From<ProjectionTarget> for TargetId {
    fn from(target: ProjectionTarget) -> Self {
        match target {
            ProjectionTarget::ClaudeCode => TargetId::ClaudeCode,
            ProjectionTarget::Codex => TargetId::Codex,
            ProjectionTarget::OpenCode => TargetId::OpenCode,
        }
    }
}

impl TryFrom<TargetId> for ProjectionTarget {
    type Error = ();

    fn try_from(target: TargetId) -> Result<Self, Self::Error> {
        match target {
            TargetId::ClaudeCode => Ok(ProjectionTarget::ClaudeCode),
            TargetId::Codex => Ok(ProjectionTarget::Codex),
            TargetId::OpenCode => Ok(ProjectionTarget::OpenCode),
            TargetId::Session => Err(()),
        }
    }
}

/// MCP transport of a registry entry.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "kebab-case")]
pub enum TransportKind {
    Stdio,
    Http,
    Sse,
}

impl TransportKind {
    pub fn as_str(self) -> &'static str {
        match self {
            TransportKind::Stdio => "stdio",
            TransportKind::Http => "http",
            TransportKind::Sse => "sse",
        }
    }
}

/// Registry file location an entry came from.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Type)]
#[serde(rename_all = "kebab-case")]
pub enum Scope {
    Global,
    Workspace,
}

impl<'de> Deserialize<'de> for Scope {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        match Scope::parse(&s) {
            Some(scope) => Ok(scope),
            None => Err(serde::de::Error::custom(format!("unknown scope: {s}"))),
        }
    }
}

impl Scope {
    pub fn as_str(self) -> &'static str {
        match self {
            Scope::Global => "global",
            Scope::Workspace => "workspace",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "global" => Some(Scope::Global),
            "workspace" | "project" => Some(Scope::Workspace),
            _ => None,
        }
    }
}

/// A registry value: a literal string or a keychain reference.
///
/// Serializes the canonical `{"secretRef":"keychain:tethys/<name>"}` shape.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(untagged)]
pub enum RegistryValue {
    Plain(String),
    Secret {
        #[serde(rename = "secretRef")]
        secret_ref: String,
    },
}

impl RegistryValue {
    pub fn plain(value: impl Into<String>) -> Self {
        Self::Plain(value.into())
    }

    pub fn secret(secret_ref: impl Into<String>) -> Self {
        Self::Secret {
            secret_ref: secret_ref.into(),
        }
    }

    pub fn as_plain(&self) -> Option<&str> {
        match self {
            Self::Plain(value) => Some(value),
            Self::Secret { .. } => None,
        }
    }

    pub fn secret_ref(&self) -> Option<&str> {
        match self {
            Self::Plain(_) => None,
            Self::Secret { secret_ref } => Some(secret_ref),
        }
    }
}

/// Per-entry Tethys metadata (`x-tethys`).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Type)]
pub struct EntryMeta {
    #[serde(default)]
    pub scope: Option<Scope>,
    #[serde(default)]
    pub providers: Option<Vec<String>>,
    #[serde(default, skip_serializing)]
    pub targets: Option<Vec<TargetId>>,
    #[serde(default = "default_enabled")]
    pub enabled: bool,
    #[serde(default)]
    pub legacy: bool,
}

#[derive(Deserialize)]
struct EntryMetaRaw {
    #[serde(default)]
    scope: Option<Scope>,
    #[serde(default)]
    providers: Option<Vec<String>>,
    #[serde(default)]
    targets: Option<Vec<TargetId>>,
    #[serde(default = "default_enabled")]
    enabled: bool,
    #[serde(default)]
    legacy: bool,
}

impl<'de> Deserialize<'de> for EntryMeta {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let raw = EntryMetaRaw::deserialize(deserializer)?;
        Ok(raw.into())
    }
}

impl From<EntryMetaRaw> for EntryMeta {
    fn from(raw: EntryMetaRaw) -> Self {
        let providers = match (raw.providers, raw.targets.as_ref()) {
            (Some(p), _) => Some(p),
            (None, Some(targets)) => {
                let p: Vec<String> = targets
                    .iter()
                    .filter(|t| **t != TargetId::Session)
                    .map(|t| t.as_str().to_string())
                    .collect();
                if p.is_empty() {
                    None
                } else {
                    Some(p)
                }
            }
            (None, None) => None,
        };
        Self {
            scope: raw.scope,
            providers,
            targets: raw.targets,
            enabled: raw.enabled,
            legacy: raw.legacy,
        }
    }
}

fn default_enabled() -> bool {
    true
}

impl Default for EntryMeta {
    fn default() -> Self {
        Self {
            scope: None,
            providers: None,
            targets: None,
            enabled: true,
            legacy: false,
        }
    }
}

impl EntryMeta {
    /// Whether this entry is allowed for `provider_id`.
    pub fn allows_provider(&self, provider_id: &str) -> bool {
        match &self.providers {
            Some(providers) => providers.iter().any(|p| p == provider_id),
            None => match &self.targets {
                Some(targets) => {
                    let non_session: Vec<&str> = targets
                        .iter()
                        .filter(|t| **t != TargetId::Session)
                        .map(|t| t.as_str())
                        .collect();
                    if non_session.is_empty() {
                        true
                    } else {
                        non_session.contains(&provider_id)
                    }
                }
                None => true,
            },
        }
    }

    /// Whether this entry should be projected to `target`.
    pub fn allows(&self, target: TargetId) -> bool {
        if let Some(targets) = &self.targets {
            return targets.contains(&target);
        }
        if target == TargetId::Session {
            true
        } else {
            self.allows_provider(target.as_str())
        }
    }
}

/// One canonical MCP registry entry.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct RegistryEntry {
    #[serde(rename = "type")]
    pub transport: TransportKind,
    #[serde(default)]
    pub command: Option<String>,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(default)]
    pub env: BTreeMap<String, RegistryValue>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub headers: BTreeMap<String, RegistryValue>,
    #[serde(rename = "x-tethys", default)]
    pub meta: EntryMeta,
}

/// Transports an agent advertised for a session.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct McpTransports {
    pub stdio: bool,
    pub http: bool,
    pub sse: bool,
}

impl McpTransports {
    pub fn all() -> Self {
        Self {
            stdio: true,
            http: true,
            sse: true,
        }
    }
}

/// One MCP server handed to an agent session.
///
/// Values are `Secret` until `resolve_secrets` runs at spawn; a resolved
/// server carries only `Plain` values and is never written to disk.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct SessionServer {
    pub name: String,
    pub transport: TransportKind,
    #[serde(default)]
    pub command: Option<String>,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(default)]
    pub env: BTreeMap<String, RegistryValue>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub headers: BTreeMap<String, RegistryValue>,
}

/// Per-entry projection state in a plan.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "kebab-case")]
pub enum EntryState {
    Pending,
    InSync,
    Drifted,
    Conflict,
    Unsupported,
}

/// One entry's state inside a projection plan.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct EntryProjection {
    pub name: String,
    pub state: EntryState,
    /// Whether this entry is part of the proposed file content.
    pub projected: bool,
    #[serde(default)]
    pub notes: Vec<String>,
}

/// Preview of one projected registry into one target file.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct ProjectionPlan {
    pub target: TargetId,
    pub path: String,
    pub scope: Scope,
    pub base_hash: String,
    pub created: bool,
    pub diff: String,
    pub content: String,
    pub entries: Vec<EntryProjection>,
    #[serde(default)]
    pub providers: Vec<String>,
}

/// Manifest record for one applied projection.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct Applied {
    pub target: TargetId,
    pub path: String,
    pub scope: Scope,
    pub file_hash: String,
    pub created: bool,
    #[serde(default)]
    pub entries: BTreeMap<String, String>,
}

/// Result of verifying a projected file against its manifest entry.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "kebab-case")]
pub enum VerifyStatus {
    InSync,
    Drifted,
    Missing,
}

/// One MCP server detected in a foreign config file.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct ImportCandidate {
    pub name: String,
    pub entry: RegistryEntry,
    pub source_path: String,
    pub scope: Scope,
    pub conflict: bool,
}

/// One source file that could not be scanned.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct ImportFailure {
    pub source_path: String,
    pub message: String,
}

/// Read-only scan across every detected tool config.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct ImportScan {
    pub candidates: Vec<ImportCandidate>,
    pub failures: Vec<ImportFailure>,
}

/// Where a skill came from.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "kebab-case")]
pub enum SkillOrigin {
    Folder,
    Archive,
    GitHub,
    Lockfile,
}

/// Skill provenance (`skills_state.source` JSON).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct SkillSource {
    pub origin: SkillOrigin,
    #[serde(default)]
    pub repo: Option<String>,
    #[serde(default)]
    pub reference: Option<String>,
    #[serde(default)]
    pub subdir: Option<String>,
    #[serde(default)]
    pub lock_hash: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
}

/// One skill in the canonical home.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct SkillInfo {
    pub name: String,
    pub scope: Scope,
    pub path: String,
    pub source: SkillSource,
    pub enabled: bool,
    pub requires_trust: bool,
    pub trusted: bool,
    pub content_hash: String,
    #[serde(default)]
    pub pinned_sha: Option<String>,
}

/// One registry entry with its scope and name, for the UI.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct RegistryEntryView {
    pub name: String,
    pub scope: Scope,
    pub entry: RegistryEntry,
}

/// Where `skills.import` should load from.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum SkillImportSource {
    Folder { path: String },
    Archive { path: String },
    GitHub { spec: String },
}

/// Cheap upstream check for one skill.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct SkillUpdateCheck {
    pub name: String,
    #[serde(default)]
    pub pinned_sha: Option<String>,
    #[serde(default)]
    pub upstream_sha: Option<String>,
    pub update_available: bool,
    #[serde(default)]
    pub error: Option<String>,
}

/// Download-and-diff preview for one skill update.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct SkillUpdatePlan {
    pub name: String,
    pub upstream_sha: String,
    pub changed_files: Vec<String>,
    pub diff: String,
}

/// Result of applying one skill update.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct SkillUpdateApplied {
    pub name: String,
    pub pinned_sha: String,
    pub content_hash: String,
}

/// A row in the attachment grid representing an MCP server.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct ServerRow {
    pub name: String,
    pub transport: TransportKind,
    pub scope: Scope,
}

/// A column in the attachment grid representing an ACP Provider.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct ProviderColumn {
    pub id: String,
    pub name: String,
    pub connected: bool,
    #[serde(default)]
    pub target: Option<ProjectionTarget>,
}

/// One cell in the attachment grid.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct AttachmentCell {
    pub server_name: String,
    pub provider_id: String,
    pub state: AttachmentState,
}

/// Attachment state of an MCP server to a Provider.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum AttachmentState {
    Attached,
    UnsupportedTransport {
        needs: TransportKind,
    },
    FileProjection {
        target: ProjectionTarget,
        state: EntryState,
    },
    Excluded,
    NotNegotiated,
}

/// The Servers × Providers attachment grid.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
pub struct AttachmentGrid {
    pub servers: Vec<ServerRow>,
    pub providers: Vec<ProviderColumn>,
    pub cells: Vec<AttachmentCell>,
}

File: packages/state/src/sync.ts
//! Sync & skills domain state (M1.11, worktree D).
//!
//! Pure selectors and view mappers over the M1.4 sync engine's wire types.
//! `mcp.attachments` is authoritative: nothing here re-derives an attachment
//! state from raw capabilities — the mappers only translate a returned
//! `AttachmentState` into a view descriptor.

import type {
  Applied,
  AttachmentCell,
  AttachmentGrid,
  AttachmentState,
  EntryState,
  ImportCandidate,
  ImportScan,
  ProjectionPlan,
  ProviderColumn,
  RegistryEntryView,
  RegistryValue,
  Scope,
  SkillImportSource,
  SkillInfo,
  SkillUpdateApplied,
  SkillUpdateCheck,
  SkillUpdatePlan,
  TargetId,
  TransportKind,
  VerifyStatus,
  WorkspaceId,
} from "@tethys/bindings";
import { useState } from "react";

/**
 * The narrow slice of the typed client the MCP surface uses. `TethysClient`
 * satisfies it; tests provide fakes. Real arguments, no `call<void>`.
 */
export interface McpSyncClient {
  mcp: {
    attachments(workspaceId: string): Promise<AttachmentGrid>;
    registry_list(workspaceId?: string): Promise<RegistryEntryView[]>;
    projection_plan(
      workspaceId: string,
      target: TargetId,
      scope: Scope,
    ): Promise<ProjectionPlan>;
    projection_apply(
      workspaceId: string,
      target: TargetId,
      scope: Scope,
      plan: ProjectionPlan,
    ): Promise<Applied>;
    projection_rollback(
      workspaceId: string,
      target: TargetId,
      scope: Scope,
    ): Promise<void>;
    projection_verify(
      workspaceId: string,
      target: TargetId,
      scope: Scope,
    ): Promise<VerifyStatus>;
    import_scan(workspaceId: string): Promise<ImportScan>;
    import_apply(
      workspaceId: string,
      candidates: ImportCandidate[],
      scope: Scope,
    ): Promise<string[]>;
  };
}

/** The narrow slice of the typed client the Skills surface uses. */
export interface SkillsClient {
  skills: {
    list(workspaceId: string): Promise<SkillInfo[]>;
    trust(workspaceId: string, scope: Scope, name: string): Promise<SkillInfo>;
    enable(
      workspaceId: string,
      scope: Scope,
      name: string,
      enabled: boolean,
    ): Promise<SkillInfo>;
    import(
      workspaceId: string,
      scope: Scope,
      source: SkillImportSource,
    ): Promise<SkillInfo>;
    update_check(
      workspaceId: string,
      scope: Scope,
      name: string,
    ): Promise<SkillUpdateCheck>;
    update_plan(
      workspaceId: string,
      scope: Scope,
      name: string,
    ): Promise<SkillUpdatePlan>;
    update_apply(
      workspaceId: string,
      scope: Scope,
      name: string,
    ): Promise<SkillUpdateApplied>;
  };
}

/** One server row with its cells resolved in Provider column order. */
export interface GridRowView {
  server: AttachmentGrid["servers"][number];
  cells: Array<{ provider: ProviderColumn; state: AttachmentState }>;
}

/** The grid's ordered rows, each carrying one state per Provider column. */
export function gridRows(grid: AttachmentGrid): GridRowView[] {
  const cellsByServer = new Map<string, Map<string, AttachmentCell>>();
  for (const cell of grid.cells) {
    let byProvider = cellsByServer.get(cell.server_name);
    if (byProvider === undefined) {
      byProvider = new Map();
      cellsByServer.set(cell.server_name, byProvider);
    }
    byProvider.set(cell.provider_id, cell);
  }

  return grid.servers.map((server) => ({
    server,
    cells: grid.providers.map((provider) => ({
      provider,
      // ponytail: engine returns the full grid; a missing cell displays as the
      // muted "not connected" state rather than being re-derived.
      state: cellsByServer.get(server.name)?.get(provider.id)?.state ?? {
        kind: "not-negotiated",
      },
    })),
  }));
}

/** Counts for `aria-rowcount` / `aria-colcount` and for choosing an empty state. */
export function gridDimensions(grid: AttachmentGrid): {
  rowCount: number;
  columnCount: number;
} {
  return { rowCount: grid.servers.length, columnCount: grid.providers.length };
}

/** Number of cells the engine reports as attached (excludes excluded/other). */
export function countAttachedCells(grid: AttachmentGrid): number {
  return grid.cells.filter((cell) => cell.state.kind === "attached").length;
}

export type CellTone = "success" | "warning" | "danger" | "muted";

/** A cell's presentation: never colour alone — every state carries a label. */
export interface CellView {
  label: string;
  tone: CellTone;
  sub: string | null;
  /** True when activating the cell opens the fallback projection surface. */
  actionable: boolean;
}

/** Exhaustiveness guard: a new `AttachmentState` variant is a compile error. */
const CELL_TONES = {
  attached: "success",
  "unsupported-transport": "warning",
  "file-projection": "warning",
  excluded: "muted",
  "not-negotiated": "muted",
} satisfies Record<AttachmentState["kind"], CellTone>;

const PROJECTION_STATE_LABELS: Record<EntryState, string> = {
  pending: "pending",
  "in-sync": "in sync",
  drifted: "drifted",
  conflict: "conflict",
  unsupported: "unsupported",
};

const PROJECTION_STATE_TONES: Record<EntryState, CellTone> = {
  pending: "muted",
  "in-sync": "success",
  drifted: "warning",
  conflict: "danger",
  unsupported: "muted",
};

/** Maps one returned `AttachmentState` to its view descriptor. */
export function renderCellState(state: AttachmentState): CellView {
  switch (state.kind) {
    case "attached":
      return {
        label: "Attached",
        tone: CELL_TONES.attached,
        sub: null,
        actionable: false,
      };
    case "unsupported-transport":
      return {
        label: "Unsupported transport",
        tone: CELL_TONES["unsupported-transport"],
        sub: `needs ${transportLabel(state.needs)}`,
        actionable: false,
      };
    case "file-projection":
      return {
        label: "File projection",
        tone: PROJECTION_STATE_TONES[state.state],
        sub: PROJECTION_STATE_LABELS[state.state],
        actionable: true,
      };
    case "excluded":
      return {
        label: "—",
        tone: CELL_TONES.excluded,
        sub: null,
        actionable: false,
      };
    case "not-negotiated":
      return {
        label: "not connected",
        tone: CELL_TONES["not-negotiated"],
        sub: null,
        actionable: false,
      };
    default: {
      const _exhaustive: never = state;
      return _exhaustive;
    }
  }
}

function transportLabel(transport: TransportKind): string {
  return transport;
}

/** True when at least one cell is on the vendor-file projection fallback. */
export function hasFallbackProjection(grid: AttachmentGrid): boolean {
  return grid.cells.some((cell) => cell.state.kind === "file-projection");
}

// --- Fixtures -------------------------------------------------------------

const fixtureCell = (
  serverName: string,
  providerId: string,
  state: AttachmentState,
): AttachmentCell => ({
  server_name: serverName,
  provider_id: providerId,
  state,
});

const FIXTURE_PROVIDERS: ProviderColumn[] = [
  { id: "claude-code", name: "Claude Code", connected: true, target: null },
  { id: "opencode", name: "OpenCode", connected: true, target: null },
  { id: "codex", name: "Codex CLI", connected: true, target: "codex" },
  { id: "kiro", name: "Kiro", connected: false, target: null },
];

/** Canonical grids exercising every `AttachmentState` variant. */
export const attachmentGridFixtures = {
  /** Every one of the five cell states, across a 3×4 matrix. */
  allStates: {
    servers: [
      { name: "context7", transport: "stdio", scope: "global" },
      { name: "linear", transport: "http", scope: "workspace" },
      { name: "sentry", transport: "sse", scope: "global" },
    ] satisfies AttachmentGrid["servers"],
    providers: FIXTURE_PROVIDERS,
    cells: [
      fixtureCell("context7", "claude-code", { kind: "attached" }),
      fixtureCell("context7", "opencode", { kind: "attached" }),
      fixtureCell("context7", "codex", {
        kind: "file-projection",
        target: "codex",
        state: "drifted",
      }),
      fixtureCell("context7", "kiro", { kind: "not-negotiated" }),
      fixtureCell("linear", "claude-code", {
        kind: "unsupported-transport",
        needs: "http",
      }),
      fixtureCell("linear", "opencode", { kind: "attached" }),
      fixtureCell("linear", "codex", {
        kind: "file-projection",
        target: "codex",
        state: "conflict",
      }),
      fixtureCell("linear", "kiro", { kind: "excluded" }),
      fixtureCell("sentry", "claude-code", {
        kind: "unsupported-transport",
        needs: "sse",
      }),
      fixtureCell("sentry", "opencode", { kind: "excluded" }),
      fixtureCell("sentry", "codex", {
        kind: "file-projection",
        target: "codex",
        state: "pending",
      }),
      fixtureCell("sentry", "kiro", { kind: "not-negotiated" }),
    ],
  } satisfies AttachmentGrid,

  /** Two servers by four Providers, all attached — the U1 row-order case. */
  twoByFour: {
    servers: [
      { name: "context7", transport: "stdio", scope: "global" },
      { name: "linear", transport: "http", scope: "workspace" },
    ],
    providers: FIXTURE_PROVIDERS,
    cells: FIXTURE_PROVIDERS.flatMap((provider) => [
      fixtureCell("context7", provider.id, { kind: "attached" }),
      fixtureCell("linear", provider.id, { kind: "attached" }),
    ]),
  } satisfies AttachmentGrid,

  /** Zero servers: the header row still renders. */
  noServers: {
    servers: [],
    providers: FIXTURE_PROVIDERS,
    cells: [],
  } satisfies AttachmentGrid,

  /** Zero Providers: the frozen server column still renders. */
  noProviders: {
    servers: [
      { name: "context7", transport: "stdio", scope: "global" },
      { name: "linear", transport: "http", scope: "workspace" },
    ],
    providers: [],
    cells: [],
  } satisfies AttachmentGrid,

  /** Both axes empty: the zero-Providers state applies. */
  empty: { servers: [], providers: [], cells: [] } satisfies AttachmentGrid,
};

/** A projection preview plan; includes the `providers` list the dialog names. */
export const projectionPlanFixture: ProjectionPlan = {
  target: "codex",
  path: "/home/user/Code/tethys/.codex/config.toml",
  scope: "workspace",
  base_hash: "b3-0123456789abcdef",
  created: false,
  diff: [
    "--- a/.codex/config.toml",
    "+++ b/.codex/config.toml",
    "@@ -1,3 +1,5 @@",
    " [mcp_servers.context7]",
    ' command = "npx"',
    '-args = ["-y", "@upstash/context7-mcp"]',
    '+args = ["-y", "@upstash/context7-mcp@latest"]',
  ].join("\n"),
  content: '[mcp_servers.context7]\ncommand = "npx"\n',
  entries: [
    { name: "context7", state: "drifted", projected: true, notes: [] },
    { name: "linear", state: "pending", projected: true, notes: [] },
  ],
  providers: ["codex"],
};

/** A foreign-config scan fixture: two candidates, one conflict, one failure. */
export const importScanFixture: ImportScan = {
  candidates: [
    {
      name: "context7",
      entry: { type: "stdio", command: "npx", args: ["-y", "context7-mcp"] },
      source_path: "/home/user/.cursor/mcp.json",
      scope: "workspace",
      conflict: false,
    },
    {
      name: "linear",
      entry: { type: "http", url: "https://mcp.linear.app/sse" },
      source_path: "/home/user/.cursor/mcp.json",
      scope: "workspace",
      conflict: true,
    },
  ],
  failures: [
    { source_path: "/home/user/.vscode/mcp.json", message: "unreadable JSON" },
  ],
};

/** Registry entry views for the Server Configuration Card. */
export const registryEntryFixtures: RegistryEntryView[] = [
  {
    name: "context7",
    scope: "global",
    entry: {
      type: "stdio",
      command: "npx",
      args: ["-y", "@upstash/context7-mcp"],
      env: {
        CONTEXT7_API_KEY: { secretRef: "keychain:tethys/context7" },
        LOG_LEVEL: "debug",
      },
    },
  },
  {
    name: "linear",
    scope: "workspace",
    entry: {
      type: "http",
      url: "https://mcp.linear.app/sse",
      headers: {
        Authorization: { secretRef: "keychain:tethys/linear" },
        "X-Client": "tethys",
      },
    },
  },
];

/** Skills for `settings.skills` list fixtures. */
export const skillInfoFixtures: SkillInfo[] = [
  {
    name: "find-docs",
    scope: "workspace",
    path: ".agents/skills/find-docs",
    source: { origin: "folder", url: null, repo: null, reference: null },
    enabled: true,
    requires_trust: false,
    trusted: true,
    content_hash: "b3-aaaa",
    pinned_sha: null,
  },
  {
    name: "audit-code",
    scope: "workspace",
    path: ".agents/skills/audit-code",
    source: { origin: "git-hub", repo: "acme/audit", reference: "v1" },
    enabled: true,
    requires_trust: true,
    trusted: false,
    content_hash: "b3-bbbb",
    pinned_sha: "abcdef12",
  },
];

// --- Workspace scope ------------------------------------------------------

/** Minimal workspace option until worktree F's typed `workspace.list` lands. */
export interface WorkspaceOption {
  id: WorkspaceId;
  name: string;
}

export const workspaceOptionFixtures: WorkspaceOption[] = [
  { id: "tethys", name: "tethys" },
  { id: "sandbox", name: "sandbox" },
];

export interface McpWorkspaceScope {
  workspaceId: WorkspaceId;
  workspaces: WorkspaceOption[];
  selectWorkspace: (id: WorkspaceId) => void;
}

/**
 * D12 fixture-backed scope hook. The real source is worktree F's trust-filtered
 * `workspace.list` (still `call<void>()`); the hook's shape does not change.
 */
export function useMcpWorkspaceScope(
  workspaces: WorkspaceOption[] = workspaceOptionFixtures,
): McpWorkspaceScope {
  const [workspaceId, setWorkspaceId] = useState<WorkspaceId>(
    workspaces[0]?.id ?? "",
  );
  return { workspaceId, workspaces, selectWorkspace: setWorkspaceId };
}

/** Renders a `keychain:…` ref for a secret; never a value. */
export function registryValueText(value: RegistryValue): {
  text: string;
  secret: boolean;
} {
  if (typeof value === "string") {
    return { text: value, secret: false };
  }
  return { text: value.secretRef, secret: true };
}
